### PR TITLE
Get and set active sprint

### PR DESCRIPTION
**Pull request by duanxianbo**
(task:15606)

image

### CODE REVIEW

**Comment by duanxianbo** [comment]
@magicoder10 we can review this one too

**Comment by naturetime10** [comment]
May be use this icon: star_rate

**Comment by naturetime10** [comment]
Let's add more top and bottom paddings.

File: src/client/sprint.client.ts [hunk 1]
```diff
@@ -38,13 +41,19 @@ export class SprintClient {
                                 }
                                 owningTeam {
                                     id
+                                    activeSprint {
+                                        id
+                                    }
```
Comment (naturetime10) [inline src/client/sprint.client.ts]:
```suggestion
```

File: src/client/sprint.client.ts [hunk 2]
```diff
@@ -38,13 +41,19 @@ export class SprintClient {
...
                                 }
                                 awaitForTasks {
                                     id
                                     status
                                     isPlanned
                                     owningTeam {
                                         id
+                                        activeSprint {
+                                            id
+                                        }
```
Comment (naturetime10) [inline src/client/sprint.client.ts]:
```suggestion
```

File: src/client/sprint.client.ts [hunk 3]
```diff
@@ -97,6 +106,9 @@ export class SprintClient {
                             endAt
                             owningTeam {
                                 id
+                                activeSprint {
+                                    id
+                                }
```
Comment (naturetime10) [inline src/client/sprint.client.ts]:
```suggestion
```

File: src/client/task.client.ts [hunk 4]
```diff
@@ -32,6 +32,9 @@ export class TaskClient {
                             effort
                             owningTeam {
                                 id
+                                activeSprint {
+                                    id
+                                }
```
Comment (naturetime10) [inline src/client/task.client.ts]:
```suggestion
```

File: src/client/task.client.ts [hunk 5]
```diff
@@ -99,6 +102,9 @@ export class TaskClient {
                             }
                             owningTeam {
                                 id
+                                activeSprint {
+                                    id
+                                }
```
Comment (naturetime10) [inline src/client/task.client.ts]:
```suggestion
```

File: src/client/task.client.ts [hunk 6]
```diff
@@ -134,6 +140,9 @@ export class TaskClient {
                             effort
                             owningTeam {
                                 id
+                                activeSprint {
+                                    id
+                                }
```
Comment (naturetime10) [inline src/client/task.client.ts]:
```suggestion
```

File: src/client/team.client.ts [hunk 7]
```diff
@@ -42,6 +45,9 @@ export class TeamClient {
                                 endAt
                                 owningTeam {
                                     id
+                                    activeSprint {
+                                        id
+                                    }
```
Comment (naturetime10) [inline src/client/team.client.ts]:
```suggestion
```

File: src/client/team.client.ts [hunk 8]
```diff
@@ -79,13 +85,19 @@ export class TeamClient {
                                 }
                                 owningTeam {
                                     id
+                                    activeSprint {
+                                        id
+                                    }
```
Comment (naturetime10) [inline src/client/team.client.ts]:
```suggestion
```

File: src/client/team.client.ts [hunk 9]
```diff
@@ -79,13 +85,19 @@ export class TeamClient {
...
                                 }
                                 awaitForTasks {
                                     id
                                     status
                                     isPlanned
                                     owningTeam {
                                         id
+                                        activeSprint {
+                                            id
+                                        }
```
Comment (naturetime10) [inline src/client/team.client.ts]:
```suggestion
```

File: src/client/team.client.ts [hunk 10]
```diff
@@ -161,6 +179,40 @@ export class TeamClient {
...
+                `,
+                variables: {
+                    teamId,
+                    sprintId,
+                },
+            })
+            .then((payload: any) => payload?.updateTeamActiveSprint);
+    }
+
+
```
Comment (naturetime10) [inline src/client/team.client.ts]:
```suggestion
```

File: src/components/App.component.module.scss [hunk 11]
```diff
@@ -20,6 +20,24 @@
         overflow-x: hidden;
         --tab-text-color: #{$secondaryColor};
 
+        .BannerContent {
+            display: flex;
+            align-items: center;
+            margin-left: auto;
+            margin-right: auto;
+
```
Comment (naturetime10) [inline src/components/App.component.module.scss]:
```suggestion
```

File: src/components/App.component.module.scss [hunk 12]
```diff
@@ -20,6 +20,24 @@
...
+            align-items: center;
+            margin-left: auto;
+            margin-right: auto;
+            
+
+            .BannerText {
+                font-size: 20px;
+                font-weight: 600;
+            }
+            .BannerAction {
```
Comment (naturetime10) [inline src/components/App.component.module.scss]:
```suggestion
            }
            
            .BannerLink {
```

File: src/components/App.component.tsx [hunk 13]
```diff
@@ -555,11 +592,19 @@ export class AppComponent extends Component<Props, State> {
...
+
         this.setState({
             me: currentUser,
             currentTeam: currTeam,
             teams: currentUser?.teams || [],
             activeTeamId: currTeam?.id,
+            showActiveSprintBanner: !!(
+                activeSprint &&
+                currentSprintId &&
+                activeSprint.id !== currentSprintId
```
Comment (naturetime10) [inline src/components/App.component.tsx]: Let's make this a separate variable.

File: src/libs/ui/Banner.module.scss [hunk 14]
```diff
@@ -0,0 +1,14 @@
+.Banner {
+    background-color: white;
+    padding: 8px 16px;
+    display: flex;
+    align-items: center;
+    justify-content: space-between;
+
```
Comment (naturetime10) [inline src/libs/ui/Banner.module.scss]:
```suggestion
```

File: src/libs/ui/Banner.tsx [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+                {this.props.closable && (
+                    <div className={styles.Close} onClick={this.handleClose}>
+                        <MaterialIconUI>close</MaterialIconUI>
+                    </div>
+                )}
+            </div>
+        );
+    }
+
+    private handleClose = () => {
```
Comment (naturetime10) [inline src/libs/ui/Banner.tsx]:
```suggestion
    private onCloseClick = () => {
```

File: src/storage/graph/team.node.ts [hunk 16]
```diff
@@ -90,4 +90,18 @@ export class TeamNode implements Team {
 
         return sprints;
     }
+
+
+    public get activeSprint(): SprintNode | undefined {
+        const appState = this.localStore.getState();
+
```
Comment (naturetime10) [inline src/storage/graph/team.node.ts]:
```suggestion
```

File: src/storage/syncer/mutation.ts [hunk 17]
```diff
@@ -11,7 +11,8 @@ export type CollectionType =
     | 'SprintTaskRelation'
     | 'Client'
     | 'TaskActivity'
-    | 'SprintParticipant';
+    | 'SprintParticipant'
+    | 'TeamActiveSprint';
```
Comment (naturetime10) [inline src/storage/syncer/mutation.ts]: If backend is changed to use `UpdateTeam`, then this is not needed anymore

**Review by naturetime10 — CHANGES_REQUESTED** [review]

**Comment by duanxianbo** [comment]
@magicoder10

File: src/components/App.component.tsx [hunk 18]
```diff
@@ -194,6 +197,28 @@ export class AppComponent extends Component<Props, State> {
...
+                                        <div className={styles.BannerContent}>
+                                            <div className={styles.BannerText}>
+                                                You are viewing inactive sprint.{' '}
+                                                <span
+                                                    className={
+                                                        styles.BannerLink
+                                                    }
+                                                    onClick={
+                                                        this
+                                                            .onActiveSprintSelect
```
Comment (naturetime10) [inline src/components/App.component.tsx]: OnGotoActiveSprintClick

File: src/components/App.component.tsx [hunk 19]
```diff
@@ -10,6 +10,8 @@ import { TeamyExtensionRuntime } from '@lib/ext/runtime';
 import { IdentityClient } from '@lib/identity/Identity.client';
 import { Route } from '@lib/router/Route';
 import { Router } from '@lib/router/router';
+import { BannerUI } from '@lib/ui/Banner';
+import { ButtonUI } from '@lib/ui/Button';
```
Comment (naturetime10) [inline src/components/App.component.tsx]: Are we still using this?

File: src/components/App.component.tsx [hunk 20]
```diff
@@ -555,11 +592,20 @@ export class AppComponent extends Component<Props, State> {
     private updateState() {
         const currentUser = this.graphSource.currentUser();
         const currTeam = this.graphSource.currentTeam();
+        const activeSprint = currTeam?.activeSprint;
+        const currentSprintId = this.graphSource.currentSprint()?.id;
+        const showActiveSprintBanner = !!(
+            activeSprint &&
+            currentSprintId &&
+            activeSprint.id !== currentSprintId
+        )
```
Comment (naturetime10) [inline src/components/App.component.tsx]: Can we simplify/breakdown this logic to remove double negative.

File: src/components/internal/SprintsHeaderSection.component.tsx [hunk 21]
```diff
@@ -81,7 +81,7 @@ export class SprintsHeaderSectionComponent extends Component<Props, State> {
                             onClick={this.onHeaderSectionClick}
                         >
                             <div className={styles.Icon}>
-                                <MaterialIconUI>directions_run</MaterialIconUI>
+                                <MaterialIconUI>{!this.isActiveSprint() ? 'directions_run' : 'star_rate'}</MaterialIconUI>
```
Comment (naturetime10) [inline src/components/internal/SprintsHeaderSection.component.tsx]: Can we reverse the condition and remove the negation?

Comment (naturetime10) [reply src/components/internal/SprintsHeaderSection.component.tsx]: Can we add a screenshot for this?

Comment (naturetime10) [reply src/components/internal/SprintsHeaderSection.component.tsx]: Can we show the icon on the right of sprint name?

File: src/components/internal/SprintsHeaderSection.component.tsx [hunk 22]
```diff
@@ -109,6 +109,20 @@ export class SprintsHeaderSectionComponent extends Component<Props, State> {
                         )}
                     </div>
                 </div>
+                <TooltipUI
+                    message={'Set current sprint to active sprint'}
```
Comment (naturetime10) [inline src/components/internal/SprintsHeaderSection.component.tsx]: Set as active sprint

File: src/components/internal/SprintsHeaderSection.component.tsx [hunk 23]
```diff
@@ -109,6 +109,20 @@ export class SprintsHeaderSectionComponent extends Component<Props, State> {
...
+                    position={'Bottom'}
+                >
+                    <div className={styles.SetActiveSprintSection}>
+                        <div
+                            className={styles.Button}
+                            onClick={this.onSetCurrentSprintClick}
```
Comment (naturetime10) [inline src/components/internal/SprintsHeaderSection.component.tsx]: OnSetActiveSprintClick

File: src/components/internal/SprintsHeaderSection.component.tsx [hunk 24]
```diff
@@ -195,6 +209,21 @@ export class SprintsHeaderSectionComponent extends Component<Props, State> {
         this.copyUnfinishedTasksConfirmTaskRef.current?.open();
     };
 
+    private isActiveSprint = () => {
+        const currentTeam = this.graphSource.currentTeam();
+        return currentTeam?.activeSprint && currentTeam.activeSprint.id === this.state.currentSprint?.id;
```
Comment (naturetime10) [inline src/components/internal/SprintsHeaderSection.component.tsx]: Let's break & to 2 lines

File: src/components/internal/SprintsHeaderSection.component.tsx [hunk 25]
```diff
@@ -195,6 +209,21 @@ export class SprintsHeaderSectionComponent extends Component<Props, State> {
...
+    }
+
+    private onSetCurrentSprintClick = () => {
```
Comment (naturetime10) [inline src/components/internal/SprintsHeaderSection.component.tsx]: onSetAsActiveSprintClick

File: src/components/internal/SprintsHeaderSection.component.tsx [hunk 26]
```diff
@@ -228,13 +257,14 @@ export class SprintsHeaderSectionComponent extends Component<Props, State> {
         }
 
         const currentSprint = this.graphSource.currentSprint()!;
+        const currentTeam = this.graphSource.currentTeam()
```
Comment (naturetime10) [inline src/components/internal/SprintsHeaderSection.component.tsx]: ;

File: src/libs/ui/Banner.tsx [hunk 27]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+interface State {
+    closed: boolean;
+}
+
+export class BannerUI extends Component<Props, State> {
+    constructor(props: Props) {
+        super(props);
+
```
Comment (naturetime10) [inline src/libs/ui/Banner.tsx]: Remove empty line.

File: src/storage/graph/team.node.ts [hunk 28]
```diff
@@ -90,4 +90,17 @@ export class TeamNode implements Team {
 
         return sprints;
     }
+
```
Comment (naturetime10) [inline src/storage/graph/team.node.ts:93]: Remove empty line.

File: src/storage/syncer/mutation.ts [hunk 29]
```diff
@@ -11,7 +11,7 @@ export type CollectionType =
     | 'SprintTaskRelation'
     | 'Client'
     | 'TaskActivity'
-    | 'SprintParticipant';
+    | 'SprintParticipant'
```
Comment (naturetime10) [inline src/storage/syncer/mutation.ts]: ;

File: src/storage/syncer/stateMutators.ts [hunk 30]
```diff
@@ -28,6 +28,7 @@ import { UserState } from '@core/storage/states/user.state';
 import {
     DeleteSprintParticipantPayload,
     DeleteTaskAwaitForRelationPayload,
+    UpdateTeamActiveSprintPayload,
```
Comment (naturetime10) [inline src/storage/syncer/stateMutators.ts]: Remove line

File: src/storage/syncer/stateMutators.ts [hunk 31]
```diff
@@ -628,6 +635,7 @@ export function mergeSprint(
     currAppState: AppState,
     remoteSprint: RemoteSprint,
 ): AppState {
+    debugger
```
Comment (naturetime10) [inline src/storage/syncer/stateMutators.ts]: Remove debugger

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: src/components/internal/SprintsHeaderSection.component.tsx [hunk 32]
```diff
@@ -52,6 +53,8 @@ export class SprintsHeaderSectionComponent extends Component<Props, State> {
     }
 
     render() {
+        const isActiveSprint = this.isActiveSprint();
+
```
Comment (naturetime10) [inline src/components/internal/SprintsHeaderSection.component.tsx]:
```suggestion
```

File: src/components/internal/shared.scss [hunk 33]
```diff
@@ -49,7 +49,9 @@ $sectionSpace: 10px;
         &:hover {
             background-color: white;
             --icon-color: #{$primaryColor};
-            cursor: pointer;
+            &:not(.Disabled) {
```
Comment (naturetime10) [inline src/components/internal/shared.scss:53]:
```suggestion

            &:not(.Disabled) {
```

**Review by naturetime10 — APPROVED** [review]

**Pull request merged**